### PR TITLE
RecoParticleFlow/PFProducer - merging PFBlockProducer and PFBlockAlgo

### PR DESCRIPTION
**Pull request by guitargeek**
#### PR description:

Part of the occasional PFProducer PRs to make the Particle Flow code easier to grasp.

* The PFBlockProducer was a very minimal wrapper around the PFBlockAlgo class. In auch a case it does not make sense to separate them, so the PFBlockAlgo was merged with the producer
* Moving already a lot of the code around, the whole new PFBlockProducer was formatted with Davids clang-format file shown in the last reco meeting (would be done anyway at some point)
* Remove preprocessor directives
* some spurious modernization of the code
* put the QuickUnion class into its own files, as I think it's general enough to deserve that (also clang-formatted)

Observations:
* after making the links between elements a first time, the links are stored in a hash table before going after the elements a second time and updating the links. That is not really efficient. There is definitely room for speedup and improvement in this producer! Does it take a significant portion of the reconstruction time?

#### PR validation:

CMSSW compiles and matrix tests pass.

### CODE REVIEW

**Comment by slava77** [comment]
We have a rather common pattern of light-weight EDProducer implementation, which relies on a main worker algorithm in many places of CMSSW.

What problem is solved by moving the rather complex algo completely inside the producer?

**Comment by guitargeek** [comment]
Ah I see so there is a concept behind having these "thin producers" that only wrap a single algorithm class? I did not know that but I'm happy to learn about it. Is it an advantage if the producers are lightweight and offload the heavy parts to libraries?

The problem I try to solve here is as I said the opaqueness of the particle flow code, and by putting the producer and the algo together the benefits are in particular that there are less files to keep track of (in particular no header files where you need to worry that they are used somewhere else), less arguments to pass around, and simpler data hierarchies, making future changes much simpler. But well, that's just my opinion, and if you don't agree let's not lose too much time with that :)

**Comment by slava77** [comment]
On 3/27/19 10:23 AM, Jonas Rembser wrote:
> The problem I try to solve here is as I said the opaqueness of the
> particle flow code, and by putting the producer and the algo together
> the benefits are in particular that there are less files to keep track
> of (in particular no header files where you need to worry that they are
> used somewhere else), less arguments to pass around, and simpler data
> hierarchies, making future changes much simpler. But well, that's just
> my opinion, and if you don't agree let's not lose too much time with that :)
> 

Somehow some of the above sounds like a roadmap to a monolithic code.
I thought that the goal was to move PF to a more modular design.


PFBlockAlgo was extensively updated as a part of the PF code
reorganization in LS1.
I'd suggest to consider this part as functionally done,
except for some straightforward cleanup or C++ modernization that may
have come about since 2015.
That's unless I'm missing something.

**Comment by guitargeek** [comment]
I don't agree, I think it's actually the roadmap to more modular code: the thing is if algo and producer are separated, developers will write new producers that use the original algo. Then, the algo gets adapted to also support specifics of the new producer, and voila you end up with a hairball where current and legacy algorithms are hard to disentangle, like the PFProducer or the electron producers. What I agree with is that the functionality of the block algo should be considered as done. It's indeed better organized than some other pieces of PF.

**Comment by slava77** [comment]
On 3/27/19 11:19 AM, Jonas Rembser wrote:
> he thing is if algo and producer are separated, developers will write
> new producers that use the original algo.

is this an abstract argument or do you know of a plan for an alternative
producer that will create PFBlocks ?

**Comment by guitargeek** [comment]
Fortunately this is just an abstract argument, but with more than a decade of CMS to look forward to, one never knows... Actually I have also less abstract arguments: what if we want to implement a fillDescriptions method for the block producer? If it relies on a separate algo class with it's own parameters, one would also need one further level of fillDescriptions, making the migration slightly more cumbersome. It's also easier to keep track of unused includes if everything is in one producer file. Of course I can also think of counterarguments, for example it gets harder to track the history of the BlockAlgo with git blame. This will get harder anyway with the upcoming clang-format, that's why I thought it's now the right time to shuffle around code a bit where it makes sense.

**Comment by slava77** [comment]
It looks like moving QuickUnion out of PFBlockAlgo may have a non-negligible cost.
Please check/confirm that the cost increase is tiny.

**Comment by slava77** [comment]
I see some limited benefit from having PFBlockAlgo in the same /plugins as the PFBlockProducer.
This way it is not accessible outside of the PFProducer package, which seems to be the concern expressed in https://github.com/cms-sw/cmssw/pull/26265#issuecomment-477290225.

I prefer to stay away from not obviously necessary merging of distinct files.

About the includes: are they really all needed? E.g. there is an include of PFMuonAlgo, which looks unnecessary.
There is an automated tool to check includes, but I forgot its name
 @ianna I recall that you used something

File: RecoParticleFlow/PFProducer/plugins/PFBlockProducer.cc [hunk 1]
```diff
@@ -1,76 +1,481 @@
...
+#include "RecoParticleFlow/PFClusterTools/interface/PFResolutionMap.h"
+#include "RecoParticleFlow/PFProducer/interface/BlockElementImporterBase.h"
+#include "RecoParticleFlow/PFProducer/interface/BlockElementLinkerBase.h"
+#include "RecoParticleFlow/PFProducer/interface/KDTreeLinkerBase.h"
+#include "RecoParticleFlow/PFProducer/interface/PFBlockElementSCEqual.h"
+#include "RecoParticleFlow/PFProducer/interface/PFBlockLink.h"
+#include "RecoParticleFlow/PFProducer/interface/PFMuonAlgo.h"
+#include "RecoParticleFlow/PFProducer/interface/PhotonSelectorAlgo.h"
+#include "RecoParticleFlow/PFProducer/interface/QuickUnion.h"
+#include "RecoParticleFlow/PFProducer/interface/Utils.h"
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/PFBlockProducer.cc]:
```
fatal error: RecoParticleFlow/PFProducer/interface/Utils.h: No such file or directory
```

**Comment by cbernet** [comment]
@guitargeek the reason why the PF algos were wrapped in producers are: 
- PF was originally developped as a standalone package purely based on root, before CMSSW even exists
- This allowed us to do the whole debugging and studies of PF during run 1 interactively in FWLite, out of CMSSW. We could not have done it in the full framework. 

I'm not sure it's a good strategy to tightly bind PF to the full framework. 

These are just my 2 cents, for the people who will have to debug and study PF at HL-LHC, I'm not objecting at all to this pull request.

**Comment by slava77** [comment]
for the includes, the tool is includechecker.
you can invoke it with ```chkincludes -p <package or subsystem>```

**Comment by slava77** [comment]
it turns out that a more modern way to check includes is the include-what-you-use (iwyu) tool.
```scram b -k -j 6 compile COMPILER=iwyu ```

**Comment by guitargeek** [comment]
Sorry for responding so late. Right, the problem with the missing file arose because of one of my other PRs. @slava77, what do you mean exactly with "It looks like moving QuickUnion out of PFBlockAlgo may have a non-negligible cost"? Probably I'm missing something here, but what kind of cost are you referring to? I don't know putting a class in a separate file would affect CPU time for example.

Thanks @cbernet for your comment! I completely understand your point of not binding things too tightly to the full framework., and I generally see the merit of that for many other producers in this package. The PFBlockProducer is different however, because it is really only a very thin wrapper around only one algo which seemed to have broken standalone compatibility a long time ago, with all these Linker, KDTrees and Importer plugins. It uses the plugin factory as well as the ParameterSet functionality of the full framework. I'm still no FWLite expert, but if I remember correctly from the situation with the Electron MVA, this might not work with FWLite anyway?

Thanks for the info with the `iwyu` tool, Slava! I will try it out here.

**Comment by slava77** [comment]
On 4/4/19 4:09 PM, Jonas Rembser wrote:
> what do you mean exactly with "It looks like moving QuickUnion out of
> PFBlockAlgo may have a non-negligible cost"? Probably I'm missing
> something here, but what kind of cost are you referring to? I don't know
> putting a class in a separate file would affect CPU time for example.

it looks like the old implementation is inlined.

**Comment by guitargeek** [comment]
Ah, right! Thanks, I will check if this has any effect.

**Comment by guitargeek** [comment]
The `iwyu` tool is pretty aggressive, you can see what it does with the PFBlockProducer in this PR here:

https://github.com/guitargeek/cmssw/blob/f92e7f67c705282ea874bae1e9af38f3d067e1e0/RecoParticleFlow/PFProducer/plugins/PFBlockProducer.cc

It really strips the includes down to the bare minimum. I think for a plugin without header file that might be used somewhere else, this might actually be good!

**Comment by slava77** [comment]
> It really strips the includes down to the bare minimum. I think for a plugin without header file that might be used somewhere else, this might actually be good! What do you think?

i think that this is good.
There were clearly includes of things completely irrelevant at least by now.

**Comment by guitargeek** [comment]
Ok I rebased onto master the fix the compilation problems, brought back the QuickUnion into the PFBlockProducer file and cleaned the includes. Maybe I'll reach the point where you at least consider testing it eventually :)

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by guitargeek** [comment]
Thanks for testing! To me it looks like the relval errors are not related to this PR, no?

**Comment by smuzaffar** [comment]
please test
cms-bot PR https://github.com/cms-sw/cms-bot/pull/1115 should fix this

**Comment by slava77** [comment]
I still did not change my mind that merging the algo into the producer implementation is useful.

**Comment by slava77** [comment]
> brought back the QuickUnion into the PFBlockProducer file

For my curiosity, what is the difference in performance, in case you checked?

**Comment by guitargeek** [comment]
Hi @slava77, sorry I did not have the time to do the benchmark in the last busy week...

Anyway, let's not do this PR then lose time over it, if we can't agree here. With all the PRs I occasionally do, it's only natural that some of them are controversial :)

Thanks

**Comment by slava77** [comment]
@guitargeek 
thank you for your understanding.

I would still like to see the cleanup of includes to enter the CMSSW release.
Also, simply moving the BlockAlgo to the plugins so that it's obviously exposed only to the producers is apparently a significant part of the change that you wanted to achieve with this PR.

**Pull request closed without merging**